File: src/o_loadAIF.m
function d = o_loadAIF( d, saveflag )
%O_LOADAIF This will load the AIF from the volts file (if unprocessed) or
%the AIF file (if available).
%   Detailed explanation goes here

% d - the DCE-FI datastructure. Fields critical for this function are:
%       d.dir, d.fname

% saveflag:
    %0 - attempt to load .aif, use volts but don't save as .aif
    %1 - attempt to load .aif, use volts and save if it doesn't exist (default)
    %2 - use volts and overwrite .aif if it exists

if nargin == 1
    saveflag = 1;
end
    
% first, check to see whether .aif is there
pred_file = [d.fname(1:max(strfind(d.fname,'_'))-1) '.aif'];

% if the .aif file exists, load it.

if (exist([d.dir pred_file]) == 2) && (saveflag ~= 2)
    
    % load the aif file
    M = csvread([d.dir pred_file]);
    
else %if there is no 'aif file' or if saveflag indicates it should be overwritten (opt 2).
    
    % first, check to see whether volts_file is there
    aif_path = [];
    files = dir(d.dir);
    for i = 3:length(files)
        if ~isempty(strfind(files(i).name, 'Volts'))
            aif_path = [d.dir files(i).name];
        end
    end
    
    if isempty(aif_path) %if a file couldn't be found
        %dlg "Could not locate the AIF file"
        hw1 = warndlg('Could not locate a .aif file or a Volts file','File Not Found');
    
    else %file was found with 'Volts' in it.
        filename = aif_path; 
        
        % input dialog to get physiological parameters.
        prompt = {'Enter SpO2 (%):','Enter total hemoglobin (g/dL):'};
        dlgtitle = 'Input'; dims = [1 35]; definput = {'97','14'};
        answer = inputdlg(prompt,dlgtitle,dims,definput);

        d.SaO2 = str2num(answer{1})./100; 
        d.tHb = str2num(answer{2}); %convert to num.
        wv = [804 938]; %wavelength of probe LEDs.

        % run TI_to_AIF and save
        filt_method = 'upenn'; % upenn is best, simple is faster.
        [ t, Ca ] = TI_to_AIF( filename, d.SaO2, d.tHb, wv, 0, filt_method, 0 );

        %select relevant portion of the curve
        h = imrect; 
        pos = getPosition(h);
        
        % crop to the approprate part.
        x1 = find(t>pos(1),1,'first');
        x2 = find(t>pos(3),1,'first');
        
        d_aif = interp1(t(x1:x2)-t(x1),Ca(x1:x2),d.t(:), 'linear', NaN );
        
        % if option 0 was chosen, do not save a .aif file, otherwise do.
        if saveflag ~= 0
            csvwrite([d.dir pred_file], [d.t(:) d_aif(:)]);
        end
    end
end

File: sol/o_loadAIF.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sun Dec 27 10:12:36 2020

@author: thomasusherwood
"""

import os
import csv
import numpy as np
from TI_to_AIF import TI_to_AIF
from pandas import read_csv
import matplotlib.pyplot as plt
import DCEFI
from typing import List

def o_loadAIF(d: DCEFI, saveflag: int = 1):
    """

    :param d: DCE-FI datastructure containing information about the saved file
    :param saveflag: determines if file should be saved
    :return: None
    """

    # first, check to see whether .aif is there
    pred_file = d.fname[0:max(strfind(d.fname,'_'))] + '.aif'
    
    # if the .aif file exists, load it.
    if (os.path.exists(d.dir+pred_file)) and \
        (os.path.splitext(d.dir+pred_file)[-1].lower() in \
        [".m" ".mlx" ".mlapp" ".mat" ".fig" ".txt"]) and (saveflag != 2):
        # load the aif file
        M = np.array(read_csv(d.dir+pred_file, sep=',',header=None))
        
        d.aif = M[:,1].T
    else:
        aif_path = np.array([])
        files = os.listdir(d.dir)
        for i in range(2,len(files)):
            if 'Volts' in files[i]:
                aif_path = np.append(aif_path, d.dir + files[i])
        
        if len(aif_path) == 0: # if a file couldn't be found
            # dlg "Could not locate the AIF file"
            raise FileNotFoundError("Could not locate a .aif file or a " + \
                                    "Volts file")
        else: # file was found with 'Volts' in it.
            filename = aif_path[0]

            wv = [804, 938] # wavelength of probe LEDs.
    
            # run TI_to_AIF and save
            filt_method = 'upenn' # upenn is best, simple is faster.
            (t, Ca) = TI_to_AIF(filename, wv, False, filt_method, 0)
    
            # select relevant portion of the curve
            clicks = plt.ginput(n=2, timeout=0, show_clicks=True)
            xmin = min(clicks[0][0],clicks[1][0])
            ymin = min(clicks[0][1],clicks[1][1])
            width = abs(max(clicks[0][0],clicks[1][0])-xmin)
            height = abs(max(clicks[0][1],clicks[1][1])-ymin)
            pos = np.array([xmin, ymin, width, height])
            
            # crop to the approprate part.
            x1 = find(t > pos[0],True)
            x2 = find(t > pos[0]+pos[2],True)

            t_cropped = t[x1-1:x2]-t[x1]
            t_cropped = t_cropped.reshape((len(t_cropped),1))
            Ca_cropped = Ca[x1-1:x2]
            Ca_cropped = Ca_cropped.reshape((len(Ca_cropped), 1))
            csv_matrix = np.hstack((t_cropped,Ca_cropped))
            
            # if option 0 was chosen, do not save a .aif file, otherwise do.
            if saveflag != 0:
                with open(d.dir+pred_file, 'w') as csvfile:
                    csvwriter = csv.writer(csvfile)
                    for r in range(len(csv_matrix)):
                        csvwriter.writerow(csv_matrix[r,:])
            
    
def strfind(string: str, elt: str) -> tuple:
    """
    Finds indices of a single character in a string

    :param string: string to be searched
    :param elt: single character to be found
    :return: tuple of indices where character is located
    """
    result = []
    length = len(string)
    for i in range(length):
        if elt == string[i]:
            result.append(i)
            
    return result

def find(lst: List[float], elt: float) -> int:
    """
    Finds the index of the first instance of an element in a list

    :param lst: list to be searched
    :param elt: element to be found
    :return: int index of the first instance of elt
    """
    for i in range(len(lst)):
        if elt == lst[i]:
            return i